scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/TaskState.java

```diff
@@ -54,8 +54,16 @@ public Timestamp getWindowEndTimestamp() {
         return windowEnd;
     }
 
-    public TaskState moveToNextWindow(long nextWindowSizeMs) {
-        return new TaskState(windowEnd, windowEnd.plus(nextWindowSizeMs, ChronoUnit.MILLIS), Optional.empty());
+    public TaskState moveToNextWindow(Timestamp now, long confidenceWindowSizeMs, long newQueryWindowSizeMs) {
+        Timestamp newWindowEnd = now.plus(-confidenceWindowSizeMs, ChronoUnit.MILLIS);
+
+        // Make sure that the window is at least newQueryWindowSizeMs long.
+        long windowLength = ChronoUnit.MILLIS.between(windowEnd.toDate().toInstant(), newWindowEnd.toDate().toInstant());
+        if (windowLength < newQueryWindowSizeMs) {
+            newWindowEnd = windowEnd.plus(newQueryWindowSizeMs, ChronoUnit.MILLIS);
```

Comment:
why not `windowEnd.plus(newQueryWindowSizeMs - windowLength, ChronoUnit.MILLIS)` ?